server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -1723,8 +1726,32 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
     /**
      * Returns a new {@link QueryRewriteContext} with the given {@code now} provider
      */
-    public QueryRewriteContext getRewriteContext(LongSupplier nowInMillis) {
-        return new QueryRewriteContext(parserConfig, client, nowInMillis);
+    public QueryRewriteContext getRewriteContext(LongSupplier nowInMillis, IndicesRequest indicesRequest) {
+        Supplier<Map<String, IndexMetadata>> indexMetadataMapSupplier = () -> {
+            // Skip unavailable indices
+            IndicesOptions.Builder indicesOptionsBuilder = IndicesOptions.builder(indicesRequest.indicesOptions())
+                .concreteTargetOptions(new IndicesOptions.ConcreteTargetOptions(true));
+            Index[] indices = indexNameExpressionResolver.concreteIndices(
+                clusterService.state(),
+                indicesOptionsBuilder.build(),
+                true,
+                indicesRequest.indices()
+            );
+
+            Map<String, IndexMetadata> indexMetadataMap = new HashMap<>();
+            for (Index index : indices) {
+                IndexMetadata indexMetadata = clusterService.state().metadata().index(index);
+                if (indexMetadata == null) {
+                    throw new IndexNotFoundException(index);
+                }
+
+                indexMetadataMap.put(index.getName(), indexMetadata);
+            }
+
+            return indexMetadataMap;
```

Comment:
what is the purpose of the Supplier as opposed to just carry the map around? That we lazily retrieve the map only when needed?

Reply:
Exactly. The map is necessary only for the `semantic_text` query (right now at least), so it doesn't make sense to generate it for every rewrite context.